### DISCUSSION TITLE
Name values to uppercase in post process tile hook

### DISCUSSION THREAD

**Discussion by wipfli** [onthegomap/planetiler#816] [Q&A] [answered]
I would like to do change all  `name` tag values to uppercase letters in a post process tile hook. Can you maybe point me to how this would roughly look like?

I am happy to create an example afterwards for this...

**Comment by pronitdas** [comment]
` @Override
  public List<VectorTile.Feature> postProcess(int zoom, List<VectorTile.Feature> items) {


    items = items.stream()
      .map(feature -> {
        String name = feature.attrs().getOrDefault("name", "").toString();
        if (name.length() > 20) {
          String shortName = name.substring(0, 20);
          feature.attrs().put("shortName", shortName.concat("..."));
        }
        return feature;
      })
      .collect(Collectors.toList());
    return items;


  }`
  
  Something like this? I am using the post process funtion to add a short name to the poi feature. Right now i am working on only a single langauge so this works fine. but i was thinking of moving this to langauge utils and making it more accesible.

**Reply by wipfli** [reply]
Thanks @pronitdas the example you shared looks like the postprocess hook which works on as per layer basis I think.

But there is now also the tile post process hook which works on per tile basis. #813 had something to do with it. Do you know how to use it?

**Comment by msbarry** [comment, marked as answer]
You could do something like:

```java
    for (var item : items) {
      Object name = item.attrs().get("name");
      if (name instanceof String s) { // also filters out nulls
        item.attrs().put("name", s.toUpperCase());
      }
    }
```

or with the new full-tile hook:

```java
  for (var layer : layers.values()) {
    for (var item: layer) {
      Object name = item.attrs().get("name");
      if (name instanceof String s) {
        item.attrs().put("name", s.toUpperCase());
      }
    }
  }
```

There's also a 1-liner for the get/modify/set:

```java
item.attrs().computeIfPresent("name", (k, v) -> v instanceof String s ? s.toUpperCase() : v);
```

But you could also do this client-side with maplibre's [upcase](https://maplibre.org/maplibre-style-spec/expressions/#upcase)?

```json
["upcase", ["get", "name"]]
```

**Reply by wipfli** [reply]
![image](https://github.com/onthegomap/planetiler/assets/53421382/d8fad7da-088a-44a1-a3e5-ec18c7109ca4)
worked thanks

**Reply by wipfli** [reply]
should I make a post tile example

**Reply by wipfli** [reply]
OK so for the example it could be nice to have a geometry operation involved. So say I have a layer "buildings" with building polygons and a layer "landuse" with "kind=residential" and "kind=educational". How could I make an intersection between buildings and educational landuse polygons? Thanks in advance!

**Comment by wipfli** [comment]
For the example I would like to do some geometry postprocessing. Say I have 3 rectangles in a GPKG source with a color attribute that can be red, green, or blue.

The 3 rectangles look like this:

![image](https://github.com/onthegomap/planetiler/assets/53421382/6962abfd-bf4f-4249-a521-b04b7a5bf687)

With the following code I can read the info about the rectangles in the post-tile hook:

```java
@Override
  public Map<String, List<VectorTile.Feature>> postProcessTileFeatures(TileCoord tileCoord, Map<String, List<VectorTile.Feature>> layers)
      throws GeometryException {
        for (var layer : layers.values()) {
          System.out.println(tileCoord);
          for (var item: layer) {
            System.out.println(item);
          }
        }
        return null;
  }
   ```

The output looks like this:

```
{x=33 y=22 z=6}
Feature[layer=rectangles, id=1, geometry=VectorGeometry[commands=int[11], geomType=POLYGON (3)], attrs={color=red}, group=-9223372036854775808]
Feature[layer=rectangles, id=2, geometry=VectorGeometry[commands=int[11], geomType=POLYGON (3)], attrs={color=green}, group=-9223372036854775808]
Feature[layer=rectangles, id=3, geometry=VectorGeometry[commands=int[11], geomType=POLYGON (3)], attrs={color=blue}, group=-9223372036854775808]
```

I would like to create a new layer with the intersections of the polygons. It would be a layer with 7 polygons and each polygon would state which colors it has, for example "red,blue", or "red,green,blue".

Can you give me a tip how to create these intersections and emit the new geometries with new attributes?

**Reply by msbarry** [reply]
There might be a JTS operation to do exactly this, but if not you should be able to piece it together from a series of intersection/difference operations. So first you would combine them all into a single red, green and blue geometry, then:

```java
r=red.difference(green).difference(blue) // r-g-b
rg=red.intersection(green).difference(blue) // r+g-b
rgb=red.intersection(green).intersection(bue) // r+g+b
```

then it's just a problem of factoring-out the common parts to minimize the number of operations. How many classes do you expect? Is it always red/green/blue? Or is it dynamic?

**Reply by wipfli** [reply]
Thanks @msbarry. Follow up question from slack was "How do I have access to the geometries in the first place?" and here your answer was:

> you can call `.geometry().decode()` to get the JTS geometry, then group them like this:
```java
@Override
  public Map<String, List<VectorTile.Feature>> postProcessTileFeatures(TileCoord tileCoord, Map<String, List<VectorTile.Feature>> layers)
      throws GeometryException {
        Map<String, List<Geometry>> byColor = new HashMap<>();
        for (var layer : layers.values()) {
          for (var item: layer) {
            var geometry = item.geometry().decode();
            byColor.computeIfAbsent("color", color -> new ArrayList<>()).add(geometry);
          }
        }
        return null;
  }
```
> there's a `GeoUtils.combinePolygons() ` utility to combine a list of polygons into a single multipolygon geometry

I will share the results here if I get something working.
